keep bool overloads

MartinNowak opened the pull request.

andralex commented on std/net/curl.d:
```diff
@@ -3748,7 +3748,7 @@ struct Curl
+    /// ditto
     void pause(bool pauseSending, bool pauseReceiving)
     {
         pause(cast(PauseSending)pauseSending,
```
these shouldn't be casts, should be constructor calls

MartinNowak replied:
Done

WalterBright commented:
No. Do not want overloads, either. See explanation https://github.com/D-Programming-Language/phobos/pull/3281#issuecomment-102721525

MartinNowak commented:
> No. Do not want overloads, either. See explanation #3281 (comment)

So, yes we have deprecated, but we don't use it. We have Flags, but don't use that either. We also have a GC but rather use @nogc, and avoid Exceptions at all cost.
That inconsistency, the lack of learning material guiding people and the huge amount of implementation bugs are the main reasons why people struggle to learn D.
For a seasoned programmer such details don't matter too much, but D beginners will have to learn 3 ways to do the same thing and find 5 forum topics to understand what the lastest flavor is.

WalterBright commented:
I understand you don't find my arguments compelling, this is, after all, a judgment call.

> the lack of learning material guiding people

This change makes this worse, as it obsoletes tutorials, books, articles, posts, etc.

> and the huge amount of implementation bugs are the main reasons why people struggle to learn D.

Bugs are not the issue here.

> D beginners will have to learn 3 ways to do the same thing and find 5 forum topics to understand what the lastest flavor is.

The constant churn of APIs is not helpful. It hast to stop.

andralex commented:
>So, yes we have deprecated, but we don't use it.

The standard library would be one of the lesser recommended places.

>We have Flags, but don't use that either.

We do in new code. Every long-living code has a history. It's a reality of life, and accepted for all languages.

>We also have a GC but rather use @nogc, and avoid Exceptions at all cost.

I think you know the answer to that so no need to rehash it...

>That inconsistency, the lack of learning material guiding people and the huge amount of implementation bugs are the main reasons why people struggle to learn D. For a seasoned programmer such details don't matter too much, but D beginners will have to learn 3 ways to do the same thing and find 5 forum topics to understand what the lastest flavor is.

But to me it seems like adding all these overloads that forward from Annas to Caiaphas and back don't make this any better.

I'm a hair's width more ambivalent than @WalterBright on this particular addition, but I need to stick with him. More importantly I think we all should stick together.

MartinNowak commented:
> > So, yes we have deprecated, but we don't use it.

> The standard library would be one of the lesser recommended places.

Then we should probably introduce a 4 eyes review principle for changes that deprecate something, b/c peoples judgment will always differ here. Maybe we can find a dedicated deprecation commissioner, so that this doesn't add additional stalls in our review pipeline.

> We do in new code. Every long-living code has a history. It's a reality of life, and accepted for all languages.

Sure, you have functions for backwards compatibility, but this change is about a consistent guiding by example in the standard library.

Back when we introduced lambdas, we should've replaced all string "a < b" lambdas in example code,
when we added @nogc we should've annotated all druntime and phobos functions, and when we added Flag, we should have complemented all boolean functions.
Doing this is important for two reasons. Applying a feature to a decent amount of code before releasing it reveals much better whether it's really worth adding than just reasoning about it's usefulness in an example. And being consistent with feature use means people will get a smaller language (and we're not talking about code breakage here).

We just had a small hacking session 2 days ago at our meetup and tried to write some nicer code examples for the front page. Well, as you hopefully know, it's hardly possible to write 10 lines of D code without running into some deficiency with it's own history.

While it's very trivial to explain why some functions use Flag and others use bool, it's an indication for a much bigger underlying issue. Your reluctance to address the issue puzzles me, please reconsider this in a bigger context.

BTW, I don't know any other language that supports 4 different flavors of returning an additional value from a function and 2 and a half way to pass a constant argument, each of which coming with it's own subtleties and history.

MartinNowak commented:
In case you don't know this, [Number of options and the paradox of choice](http://en.wikipedia.org/wiki/Choice#Number_of_options_and_the_paradox_of_choice).

MartinNowak commented:
> More importantly I think we all should stick together.

Sure :), but a profound dissent is healthy for a debate.

JakobOvrum commented:
The more I think about it, the more leaving in both overloads makes sense to me.

Deprecating the old overloads is a bad idea. Boolean blindness costs development time but I don't think it's *bug-prone*. If our deprecations are unlikely to fix any bugs, then they are likely to be unwelcome.

However, leaving in both overloads allows new code to be more readable, saving on development time, with a truly miniscule cost. It's easy to explain in the documentation in a single line that the `Flag` overload is to be preferred because of the issue of boolean blindness (even if phrased otherwise, a single line or so should be adequate).

This doesn't have any scalability issues either, as new code will use `Flag`.

I think there's good reason to merge this.

mrkline commented:
I'll pipe down after this since I obviously don't have as much experience or esteem as the rest of you when it comes to D, but from personal experience I can say that @MartinNowak isn't speculating here. He's hitting the nail on the head.

I was quite confused the first time I looked at the docs for `std.algorithm` and saw all these strings instead of the lambdas I had just read about in TDPL. Coworkers that I have recently convinced to use D for some internal tooling at work - many among them brilliant engineers more than a decade my senior - were even more perplexed.

I introduced #3281 (not suspecting it would cause such a hullabaloo) because the docs for `Flag` are right. I never have any issue remembering `KeepTerminator` when using `std.stdio.byLine`, but have to run off to dlang.org and look up `std.file.dirEntries` every time to remember which way the boolean argument goes. I'm sorry I didn't know more about the ongoing discussions regarding deprecation and bit rot.

I (and I assume most other newcomers) turn to Phobos to see The Right Way™ to do things. As I see it, a standard library's job is not just to provide useful solutions for common problems, but also to demonstrate what idiomatic code in its host language looks like. Reading the Phobos docs doesn't always give a sense of the consistency I would hope for when trying to learn.

And with that, I'll butt out of this conversation unless I'm addressed directly. Cheers.

andralex commented:
>And with that, I'll butt out of this conversation unless I'm addressed directly. Cheers.

Thanks. That's wise to do for us all. I'll afford one last comment before I leave as well.

We're suffering from an endemic pattern:

1. A small contribution is proposed. It has a nonzero cost and an inconsequential but nonzero value.
2. A fierce debate builds around accepting the contribution.

This has really exacerbated lately. There's been just too much of "take this code as is no matter how inconsequential/unnecessarily complex/breaking/plain mediocre/wonky it is, or else". We're wasting a lot of time debating whether we should make a step to the right or to the left, and we forget we're supposed to put a step forward. We should convert that pattern into:

1. A small contribution is proposed. It has a nonzero cost and an inconsequential but nonzero value.
2. We work with the author to convert that into a better, more impactful contribution, and future ones as well.
3. Over time an esprit of pushing the state of the art forward FOR REAL develops.

(In this case the real winner should be @mrkline.) Our current "esprit" is to fight forever around the littlest things, and let real, impactful work rust (heh) in the backyard. It's amazing. Comparable effort has been collectively put in discussing this and other silly debates around refactoring that improve nothing or languge changes that break correct code about as often as wrong code, than in reviewing code that weans Phobos of gratuitous use of the GC, which is taken straight from the vision document. 

A few possibly good outcomes from this:

>Then we should probably introduce a 4 eyes review principle for changes that deprecate something, b/c peoples judgment will always differ here. Maybe we can find a dedicated deprecation commissioner, so that this doesn't add additional stalls in our review pipeline.

I am that guy. Please send all library deprecations to me, and merge none without my explicit approval.

>Back when we introduced lambdas, we should've replaced all string "a < b" lambdas in example code,
when we added @nogc we should've annotated all druntime and phobos functions

I think these remain good work items. Bool to Flag is much ado about too little.

>Well, as you hopefully know, it's hardly possible to write 10 lines of D code without running into some deficiency with it's own history.

This is surprising to me. If you really think there is a solid point here, I'd be in your debt if you substantiated.

>While it's very trivial to explain why some functions use Flag and others use bool, it's an indication for a much bigger underlying issue. Your reluctance to address the issue puzzles me, please reconsider this in a bigger context.

What I see is an inconsequential matter that we revel in debating ad nauseam. In the meantime I have your RefCounted and Unique improvements opened somewhere in my browser, but I don't find the time to review them because I now fear the next Phobos release will break code left right and center if I turn my back for a minute. This must stop.

MartinNowak commented:
> We're wasting a lot of time debating whether we should make a step to the right or to the left, and we forget we're supposed to put a step forward.

Nothing special here, it's a well know paradox.
[The Softer the Topic, the Longer the Debate](http://producingoss.com/en/common-pitfalls.html#bikeshed)

> This is surprising to me. If you really think there is a solid point here, I'd be in your debt if you substantiated.

We had 5 teams to write a small example.

- Two teams were using std.net.curl to query some API and write a small tool.
  Both of which would obviously run into the well know curl link order [Issue 7044 – Missing a way to control the order of arguments passed to the linker makes impossible to link some programs](https://issues.dlang.org/show_bug.cgi?id=7044), needing expert help from Iain and me.
  One team were D/D2 beginners and would look at [std.net.curl](https://github.com/D-Programming-Language/phobos/blob/master/std/net/curl.d) to find out how to use that library (the word documentation doesn't occur on our front page).
- Another team was working on some range algorithm, a filter or so, and was first annoyed by not having a set container, then started to explore all the glory details of the `void[0]` hack, and soon after that of zero-sized static arrays `T[0]` in general (and yes in `struct Foo { int[0] a, b; }` `&a` and `&b` really have the same address, even though that would be illegal in C or so).
- I wrote a RPN calculator and was missing std.container.stack, because using the builtin array as stack reallocates every time. So I ended up using std.container.array and spent some time trying to make a nice pop primitive, b/c you really don't want `int pop() { auto v = stack.back; stack.removeBack; return v; }` on the frontpage of dlang.org. This and the lack of a `popBack+return` primitive are one of those trivial wrappers that curiously missing.

Not sure what Iain was doing, will come back with more details once we gathered the code.

andralex commented:
@MartinNowak: good anecdotes, and good topics to work toward! For what I can see:

* Allowing choice of link order seems an additive feature (no compatibility breakage).
* Documentation of std.net.curl and others _is_ something we can and need to improve incrementally!
* Progress with containers is gated by progress with std.allocator which is gated by my putting time in that instead of this. So in a very direct way this entire matter is of net negative value. (As an aside: an engineer may just choose an associative array of bool and call it a day.)
* Same issue with the stack container
* I'd be hard pressed to qualify the choice of pointer/length vs. pointer/length/capacity a mistake. Either choice for the built-in has pros and cons, and then for everything else you need to add user-defined types. The slimmer choice for a basic type sounds like the right choice because it can be composed with into a larger structure. Once the capacity field is in, it can't be shed - it's everywhere. At any rate, this whole thing is water under the bridge as it's quite unlikely to operate a change at that level. The right response to using a slice as a container is to offer a compelling std.container, which again is a matter of us not wasting time on Useless Debates.

I suggest you convert some of these anecdotes into work items on trello (nice initiative) or bugzilla. They're of no use as support in this conversation.

MartinNowak commented:
> I'd be hard pressed to qualify the choice of pointer/length vs. pointer/length/capacity a mistake.

It was about that the current array has an implicit capacity in the GC, which is __not__ a member, and how bad slicing interacts with array appending, not important though as we can't easily change this.

I made one ticket for a slightly controversial topic ;).
[Issue 14598 – range primitives for popFront+return front](https://issues.dlang.org/show_bug.cgi?id=14598)
And I also have a pull request for std.net.curl, but some extra features were requested during the pull https://github.com/D-Programming-Language/phobos/pull/3009.
